File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapping = {')':'(', ']':'[', '}':'{'}
        stack = []
        for char in s:
            if char in mapping.values():
                stack.append(char)
                
            elif char in mapping.keys():
                if stack == [] or mapping[char] != stack.pop():
                    return False
        return stack == []
            
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(char c:s.toCharArray()){
            if(c == '(' || c == '{' || c=='['){
                stack.push(c);
            }
            else if(c == ')' || c == '}' || c ==']'){
                if(stack.isEmpty()){
                    return false;
                }
                char top = stack.pop();
                if(!isMatching(top,c)){
                    return false;
                }
                    
            }
        }
        return stack.isEmpty();
        
    }
    private static boolean isMatching(char open, char close) {
        return (open == '(' && close == ')') ||
               (open == '{' && close == '}') ||
               (open == '[' && close == ']');
}}